File: src/dots/dots_cpp.ss/src/Parameters.cpp
/******************************************************************************
*
* Copyright Saab AB, 2006-2013 (http://safirsdkcore.com)
*
* Created by: Lars Hagström / stlrha
*
*******************************************************************************
*
* This file is part of Safir SDK Core.
*
* Safir SDK Core is free software: you can redistribute it and/or modify
* it under the terms of version 3 of the GNU General Public License as
* published by the Free Software Foundation.
*
* Safir SDK Core is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Safir SDK Core.  If not, see <http://www.gnu.org/licenses/>.
*
******************************************************************************/

#include <Safir/Dob/Typesystem/Parameters.h>
#include <Safir/Dob/Typesystem/Exceptions.h>
#include <Safir/Dob/Typesystem/Utilities.h>
#include <Safir/Dob/Typesystem/Object.h>
#include <Safir/Dob/Typesystem/ObjectFactory.h>
#include <Safir/Dob/Typesystem/Internal/Kernel.h>
#include <Safir/Dob/Typesystem/Internal/InternalOperations.h>
#include <sstream>

namespace Safir
{
namespace Dob
{
namespace Typesystem
{
namespace Parameters
{
    //Parameter handling
    Int32
    GetNumberOfParameters(const TypeId typeId)
    {
        const Int32 parameters = DotsC_GetNumberOfParameters(typeId);
        if (parameters == -1)
        {
            throw IllegalValueException(L"There is no such type defined", __WFILE__, __LINE__);
        }
        else
        {
            return parameters;
        }
    }

    ParameterIndex
    GetIndex(const TypeId typeId,
             const std::wstring& parameterName)
    {
        const ParameterIndex result = DotsC_GetParameterId(typeId, Utilities::ToUtf8(parameterName).c_str());
        if (result == -1)
        {
            const char* typeName = DotsC_GetTypeName(typeId);
            std::wostringstream ostr;

            if (typeName != nullptr)
            {
                ostr << "The class '"<< Safir::Dob::Typesystem::Utilities::ToWstring(typeName) <<
                        "' does not have a parameter called '" << parameterName << "'!";
            }
            else
            {
                ostr << L"Trying to read parameter '" << parameterName << "' from a class that does not exist. TypeId=" << typeId;
            }

            throw IllegalValueException(ostr.str(), __WFILE__, __LINE__);
        }
        else
        {
            return result;
        }
    }

    const std::wstring
    GetName(const TypeId typeId,
            const ParameterIndex parameter)
    {
        DotsC_MemberType parameterType;
        DotsC_MemberType keyType;
        const char* parameterName;
        DotsC_TypeId complexTypeId;
        DotsC_TypeId keyTypeId;
        DotsC_CollectionType collectionType;
        DotsC_Int32 numberOfValues;
        DotsC_GetParameterInfo(typeId, parameter, parameterType, keyType, parameterName, complexTypeId, keyTypeId, collectionType, numberOfValues);

        return Utilities::ToWstring(parameterName);
    }

    std::wstring GetTypeName(const Dob::Typesystem::TypeId typeId, const Dob::Typesystem::ParameterIndex parameter)
    {
        DotsC_MemberType parameterType;
        DotsC_MemberType keyType;
        const char* parameterName;
        DotsC_TypeId complexTypeId;
        DotsC_TypeId keyTypeId;
        DotsC_CollectionType collectionType;
        DotsC_Int32 numberOfValues;
        DotsC_GetParameterInfo(typeId, parameter, parameterType, keyType, parameterName, complexTypeId, keyTypeId, collectionType, numberOfValues);

        if (parameterType==ObjectMemberType || parameterType==EnumerationMemberType)
        {
            return Safir::Dob::Typesystem::Utilities::ToWstring(DotsC_GetTypeName(complexTypeId));
        }
        else
        {
            return Safir::Dob::Typesystem::Utilities::ToWstring(DotsC_MemberTypeName(parameterType));
        }
    }

    MemberType
    GetType(const TypeId typeId,
            const ParameterIndex parameter)
    {
        DotsC_MemberType parameterType;
        DotsC_MemberType keyType;
        const char* parameterName;
        DotsC_TypeId complexTypeId;
        DotsC_TypeId keyTypeId;
        DotsC_CollectionType collectionType;
        DotsC_Int32 numberOfValues;
        DotsC_GetParameterInfo(typeId, parameter, parameterType, keyType, parameterName, complexTypeId, keyTypeId, collectionType, numberOfValues);

        return parameterType;
    }

    Int32
    GetArraySize(const TypeId typeId,
                 const ParameterIndex parameter)
    {
        return GetCollectionSize(typeId, parameter);
    }

    Int32
    GetCollectionSize(const TypeId typeId,
                      const ParameterIndex parameter)
    {
        DotsC_MemberType parameterType;
        DotsC_MemberType keyType;
        const char* parameterName;
        DotsC_TypeId complexTypeId;
        DotsC_TypeId keyTypeId;
        DotsC_CollectionType collectionType;
        DotsC_Int32 numberOfValues;
        DotsC_GetParameterInfo(typeId, parameter, parameterType, keyType, parameterName, complexTypeId, keyTypeId, collectionType, numberOfValues);

        return numberOfValues;
    }

    //Get parameters
    bool
    GetBoolean(const TypeId typeId,
               const ParameterIndex parameter,
               const Dob::Typesystem::ArrayIndex index)
    {
        bool b;
        DotsC_GetBooleanParameter(typeId, parameter, index, b);
        return b;
    }

    Int32
    GetEnumeration(const TypeId typeId,
                   const ParameterIndex parameter,
                   const Dob::Typesystem::ArrayIndex index)
    {
        Int32 i;
        DotsC_GetEnumerationParameter(typeId, parameter, index, DotsC_ValueMode, i);
        return i;
    }

    Int32
    GetInt32(const TypeId typeId,
             const ParameterIndex parameter,
             const Dob::Typesystem::ArrayIndex index)
    {
        Int32 i;
        DotsC_GetInt32Parameter(typeId, parameter, index, DotsC_ValueMode, i);
        return i;
    }

    Int64
    GetInt64(const TypeId typeId,
             const ParameterIndex parameter,
             const Dob::Typesystem::ArrayIndex index)
    {
        Int64 i;
        DotsC_GetInt64Parameter(typeId, parameter, index, DotsC_ValueMode, i);
        return i;
    }

    Float32
    GetFloat32(const TypeId typeId,
               const ParameterIndex parameter,
               const Dob::Typesystem::ArrayIndex index)
    {
        Float32 i;
        DotsC_GetFloat32Parameter(typeId, parameter, index, i);
        return i;
    }

    Float64
    GetFloat64(const TypeId typeId,
               const ParameterIndex parameter,
               const Dob::Typesystem::ArrayIndex index)
    {
        Float64 i;
        DotsC_GetFloat64Parameter(typeId, parameter, index, i);
        return i;
    }

    TypeId
    GetTypeId(const TypeId typeId,
              const ParameterIndex parameter,
              const Dob::Typesystem::ArrayIndex index)
    {
        TypeId t;
        DotsC_GetTypeIdParameter(typeId, parameter, index, DotsC_ValueMode, t);
        return t;
    }

    const InstanceId
    GetInstanceId(const TypeId typeId,
                  const ParameterIndex parameter,
                  const Dob::Typesystem::ArrayIndex index)
    {
        Int64 hashVal;
        const char * strVal;
        DotsC_GetHashedIdParameter(typeId, parameter, index, DotsC_ValueMode, hashVal, strVal);
        if (strVal == NULL)
        {
            return InstanceId(hashVal);
        }
        else
        {
            return InstanceId(hashVal, Utilities::ToWstring(strVal));
        }
    }

    const EntityId
    GetEntityId(const TypeId typeId,
                const ParameterIndex parameter,
                const Dob::Typesystem::ArrayIndex index)
    {
        DotsC_EntityId eid;
        const char * instanceIdStr;
        DotsC_GetEntityIdParameter(typeId, parameter, index, DotsC_ValueMode, eid, instanceIdStr);
        if(instanceIdStr==NULL)
        {
            return EntityId(eid.typeId,InstanceId(eid.instanceId));
        }
        else
        {
            return EntityId(eid.typeId,InstanceId(eid.instanceId,Utilities::ToWstring(instanceIdStr)));
        }
    }


    const ChannelId
    GetChannelId(const TypeId typeId,
                  const ParameterIndex parameter,
                  const Dob::Typesystem::ArrayIndex index)
    {
        Int64 hashVal;
        const char * strVal;
        DotsC_GetHashedIdParameter(typeId, parameter, index, DotsC_ValueMode, hashVal, strVal);
        if (strVal == NULL)
        {
            return ChannelId(hashVal);
        }
        else
        {
            return ChannelId(hashVal,Utilities::ToWstring(strVal));
        }
    }


    const HandlerId
    GetHandlerId(const TypeId typeId,
                  const ParameterIndex parameter,
                  const Dob::Typesystem::ArrayIndex index)
    {
        Int64 hashVal;
        const char * strVal;
        DotsC_GetHashedIdParameter(typeId, parameter, index, DotsC_ValueMode, hashVal, strVal);
        if (strVal == NULL)
        {
            return HandlerId(hashVal);
        }
        else
        {
            return HandlerId(hashVal,Utilities::ToWstring(strVal));
        }
    }


    const std::wstring
    GetString(const TypeId typeId,
              const ParameterIndex parameter,
              const Dob::Typesystem::ArrayIndex index)
    {
        const char* tmp;
        DotsC_GetStringParameter(typeId, parameter, index, DotsC_ValueMode, tmp);
        return Utilities::ToWstring(tmp);
    }

    const ObjectPtr
    GetObject(const TypeId typeId,
              const ParameterIndex parameter,
              const Dob::Typesystem::ArrayIndex index)
    {
        const char * tmp;
        DotsC_GetObjectParameter(typeId, parameter, index, tmp);

        ObjectPtr obj = ObjectFactory::Instance().CreateObject(tmp);
        obj->SetChanged(false);
        return obj;
    }

    const Dob::Typesystem::Binary
    GetBinary(  const Dob::Typesystem::TypeId typeId,
                const Dob::Typesystem::ParameterIndex parameter,
                const Dob::Typesystem::ArrayIndex index)
    {
        DotsC_Int32 size;
        const char * tmp;
        DotsC_GetBinaryParameter(typeId, parameter, index, tmp, size);
        Safir::Dob::Typesystem::Binary binary(tmp, tmp+size);
        return binary;
    }

    Int32 DictionaryKeyToIndex(const TypeId typeId,
                               const ParameterIndex parameter,
                               const Int32 key)
    {
        return DotsC_DictionaryInt32KeyToIndex(typeId, parameter, key);
    }

    Int32 DictionaryKeyToIndex(const TypeId typeId,
                               const ParameterIndex parameter,
                               const Int64 key)
    {
        return DotsC_DictionaryInt64KeyToIndex(typeId, parameter, key);
    }


    Int32 DictionaryKeyToIndex(const TypeId typeId,
                               const ParameterIndex parameter,
                               const std::wstring& key)
    {
        return DotsC_DictionaryStringKeyToIndex(typeId, parameter, Utilities::ToUtf8(key).c_str());
    }


    Int32 DictionaryKeyToIndex(const TypeId typeId,
                               const ParameterIndex parameter,
                               const EntityId& key)
    {
        DotsC_EntityId eid={key.GetTypeId(), key.GetInstanceId().GetRawValue()};
        return DotsC_DictionaryEntityIdKeyToIndex(typeId, parameter, eid);
    }

    Int32 DictionaryKeyToIndex(const TypeId typeId,
                               const ParameterIndex parameter,
                               const InstanceId& key)
    {
        return DotsC_DictionaryInt64KeyToIndex(typeId, parameter, key.GetRawValue());
    }

    Int32 DictionaryKeyToIndex(const TypeId typeId,
                               const ParameterIndex parameter,
                               const HandlerId& key)
    {
        return DotsC_DictionaryInt64KeyToIndex(typeId, parameter, key.GetRawValue());
    }

    Int32 DictionaryKeyToIndex(const TypeId typeId,
                               const ParameterIndex parameter,
                               const ChannelId& key)
    {
        return DotsC_DictionaryInt64KeyToIndex(typeId, parameter, key.GetRawValue());
    }


    void GetInfo(const Dob::Typesystem::TypeId typeId,
                 const Dob::Typesystem::ParameterIndex parameter,
                 Dob::Typesystem::MemberType& parameterType,
                 Dob::Typesystem::MemberType& keyType,
                 std::wstring& parameterName,
                 Dob::Typesystem::TypeId& parameterTypeId,
                 Dob::Typesystem::TypeId& keyTypeId,
                 Dob::Typesystem::CollectionType & collectionType,
                 Dob::Typesystem::Int32& numberOfValues)
    {
        const char* parameterNameUtf8 = nullptr;

        DotsC_GetParameterInfo(typeId,
                               parameter,
                               parameterType,
                               keyType,
                               parameterNameUtf8,
                               parameterTypeId,
                               keyTypeId,
                               collectionType,
                               numberOfValues);

        if (parameterNameUtf8 != nullptr)
        {
            parameterName = Utilities::ToWstring(parameterNameUtf8);
        }
    }

    Int32
    GetEnumerationDictionaryKey(const TypeId typeId,
                                const ParameterIndex parameter,
                                const Dob::Typesystem::ArrayIndex index)
    {
        Int32 i;
        DotsC_GetEnumerationParameter(typeId, parameter, index, DotsC_KeyMode, i);
        return i;
    }

    Int32
    GetInt32DictionaryKey(const TypeId typeId,
                          const ParameterIndex parameter,
                          const Dob::Typesystem::ArrayIndex index)
    {
        Int32 i;
        DotsC_GetInt32Parameter(typeId, parameter, index, DotsC_KeyMode, i);
        return i;
    }

    Int64
    GetInt64DictionaryKey(const TypeId typeId,
                          const ParameterIndex parameter,
                          const Dob::Typesystem::ArrayIndex index)
    {
        Int64 i;
        DotsC_GetInt64Parameter(typeId, parameter, index, DotsC_KeyMode, i);
        return i;
    }


    TypeId
    GetTypeIdDictionaryKey(const TypeId typeId,
                           const ParameterIndex parameter,
                           const Dob::Typesystem::ArrayIndex index)
    {
        TypeId t;
        DotsC_GetTypeIdParameter(typeId, parameter, index, DotsC_KeyMode, t);
        return t;
    }

    const InstanceId
    GetInstanceIdDictionaryKey(const TypeId typeId,
                               const ParameterIndex parameter,
                               const Dob::Typesystem::ArrayIndex index)
    {
        Int64 hashVal;
        const char * strVal;
        DotsC_GetHashedIdParameter(typeId, parameter, index, DotsC_KeyMode, hashVal, strVal);
        if (strVal == NULL)
        {
            return InstanceId(hashVal);
        }
        else
        {
            return InstanceId(hashVal, Utilities::ToWstring(strVal));
        }
    }

    const EntityId
    GetEntityIdDictionaryKey(const TypeId typeId,
                             const ParameterIndex parameter,
                             const Dob::Typesystem::ArrayIndex index)
    {
        DotsC_EntityId eid;
        const char * instanceIdStr;
        DotsC_GetEntityIdParameter(typeId, parameter, index, DotsC_KeyMode, eid, instanceIdStr);
        if(instanceIdStr==NULL)
        {
            return EntityId(eid.typeId,InstanceId(eid.instanceId));
        }
        else
        {
            return EntityId(eid.typeId,InstanceId(eid.instanceId,Utilities::ToWstring(instanceIdStr)));
        }
    }


    const ChannelId
    GetChannelIdDictionaryKey(const TypeId typeId,
                              const ParameterIndex parameter,
                              const Dob::Typesystem::ArrayIndex index)
    {
        Int64 hashVal;
        const char * strVal;
        DotsC_GetHashedIdParameter(typeId, parameter, index, DotsC_KeyMode, hashVal, strVal);
        if (strVal == NULL)
        {
            return ChannelId(hashVal);
        }
        else
        {
            return ChannelId(hashVal,Utilities::ToWstring(strVal));
        }
    }


    const HandlerId
    GetHandlerIdDictionaryKey(const TypeId typeId,
                              const ParameterIndex parameter,
                              const Dob::Typesystem::ArrayIndex index)
    {
        Int64 hashVal;
        const char * strVal;
        DotsC_GetHashedIdParameter(typeId, parameter, index, DotsC_KeyMode, hashVal, strVal);
        if (strVal == NULL)
        {
            return HandlerId(hashVal);
        }
        else
        {
            return HandlerId(hashVal,Utilities::ToWstring(strVal));
        }
    }

    const std::wstring
    GetStringDictionaryKey(const TypeId typeId,
                           const ParameterIndex parameter,
                           const Dob::Typesystem::ArrayIndex index)
    {
        const char* tmp;
        DotsC_GetStringParameter(typeId, parameter, index, DotsC_KeyMode, tmp);
        return Utilities::ToWstring(tmp);
    }
}
}
}
}

File: 4.5.x/src/dots/dots_java.ss/src/com.saabgroup.safir.dob.typesystem/src/com/saabgroup/safir/dob/typesystem/Parameters.java
// -*- coding: utf-8 -*-
/******************************************************************************
*
* Copyright Saab AB, 2005-2013 (http://safir.sourceforge.net)
*
* Created by: Lars Hagström / stlrha
*
*******************************************************************************
*
* This file is part of Safir SDK Core.
*
* Safir SDK Core is free software: you can redistribute it and/or modify
* it under the terms of version 3 of the GNU General Public License as
* published by the Free Software Foundation.
*
* Safir SDK Core is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with Safir SDK Core.  If not, see <http://www.gnu.org/licenses/>.
*
******************************************************************************/

package com.saabgroup.safir.dob.typesystem;


/**
 * Functions for getting parameter information from types.
 *
 * With these operations you can get parameter values from types.
 * You can also get information about the parameters in a type, such as
 * parameter names and indexes, TypeIds of parameters etc.
 */
public class Parameters {

    /**
     * Get the number of parameters defined in a class.
     *
     * @param typeId [in] - TypeId of class.
     * @return The number of parameters.
     * @throws IllegalValueException There is no such type defined.
     */
    public static int getNumberOfParameters(long typeId)
    {
        int result = Kernel.GetNumberOfParameters(typeId);
        if (result == -1)
        {
            throw new IllegalValueException("There is no such type or parameter defined");
        }
        else
        {
            return result;
        }
    }

    /**
     * Gets index of a named parameter.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameterName [in] - Name of parameter.
     * @return index of the named parameter.
     * @throws IllegalValueException There is no such type or parameter defined.
     */
    public static int getIndex(long typeId, String parameterName)
    {
        int result = Kernel.GetParameterId(typeId, parameterName);
        if (result == -1)
        {
            throw new IllegalValueException("There is no such type or parameter defined");
        }
        else
        {
            return result;
        }
    }

    /**
     * Get the name of the specified parameter as it was defined in the xml description.
     *
     * If the parameter does not exist the returned value is undefined. Use
     * #GetIndex to get a valid ParameterIndex, which is guaranteed to exist.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @return The name of the parameter.
     */
    public static String getName(long typeId, int parameter)
    {
        return Kernel.GetParameterName(typeId, parameter);
    }

    /**
     * Get the type of a parameter.
     *
     * If the parameter does not exist the returned value is undefined. Use
     * #GetIndex to get a valid ParameterIndex, which is guaranteed to exist.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @return The type of the parameter.
     */
    public static MemberType getType(long typeId, int parameter)
    {
        return MemberType.values()[Kernel.GetParameterType(typeId, parameter)];
    }

    /**
     * Gets a string representation of the type of a parameter.
     *
     * If the parameter is not an object or enumeration the result is undefined.
     *
     * If the parameter does not exist the returned value is undefined. Use
     * #GetIndex to get a valid ParameterIndex, which is guaranteed to exist.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @return Name of the parameter type.
     */
    public static String getTypeName(long typeId, int parameter)
    {
        return Kernel.GetParameterTypeName(typeId, parameter);
    }

    /**
     * Get the array size of a parameter.
     *
     * If the parameter does not exist the returned value is undefined. Use
     * #GetIndex to get a valid ParameterIndex, which is guaranteed to exist.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @return The array size of the parameter, or 1 if it is not an array.
     */
    public static int getArraySize(long typeId, int parameter)
    {
        return Kernel.GetParameterArraySize(typeId, parameter);
    }


    //************************************************************************************
    //* Functions for retrieval of parameters
    //************************************************************************************

    /**
     * Get a boolean parameter value.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @param index [in] - Array index. If parameter is not an array this shall be 0.
     * @return Parameter value.
     */
    public static boolean getBoolean(long typeId, int parameter, int index)
    {
        boolean val [] = new boolean [1];
        Kernel.GetBooleanParameter(typeId, parameter, index, val);
        return val[0];
    }

    /**
     * Get an enumeration parameter value.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @param index [in] - Array index. If parameter is not an array this shall be 0.
     * @return Parameter value.
     */
    public static int getEnumeration(long typeId, int parameter, int index)
    {
        int val[] = new int[1];
        Kernel.GetInt32Parameter(typeId, parameter, index, val);
        return val[0];
    }

    /**
     * Get an Int32 parameter value.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @param index [in] - Array index. If parameter is not an array this shall be 0.
     * @return Parameter value.
     */
    public static int getInt32(long typeId, int parameter, int index)
    {
        int val [] = new int [1];
        Kernel.GetInt32Parameter(typeId, parameter, index, val);
        return val[0];
    }

    /**
     * Get an Int64 parameter value.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @param index [in] - Array index. If parameter is not an array this shall be 0.
     * @return Parameter value.
     */
    public static long getInt64(long typeId, int parameter, int index)
    {
        long val [] = new long [1];
        Kernel.GetInt64Parameter(typeId, parameter, index, val);
        return val[0];
    }

    /**
     * Get a Float32 parameter value.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @param index [in] - Array index. If parameter is not an array this shall be 0.
     * @return Parameter value.
     */
    public static float getFloat32(long typeId, int parameter, int index)
    {
        float val [] = new float [1];
        Kernel.GetFloat32Parameter(typeId, parameter, index, val);
        return val[0];
    }

    /**
     * Get a Float64 parameter value.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @param index [in] - Array index. If parameter is not an array this shall be 0.
     * @return Parameter value.
     */
    public static double getFloat64(long typeId, int parameter, int index)
    {
        double val [] = new double [1];
        Kernel.GetFloat64Parameter(typeId, parameter, index, val);
        return val[0];
    }

    /**
     * Get a string parameter value.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @param index [in] - Array index. If parameter is not an array this shall be 0.
     * @return Parameter value.
     */
    public static String getString(long typeId, int parameter, int index)
    {
        String val [] = new String [1];
        Kernel.GetStringParameter(typeId, parameter, index, val);
        return val[0];
    }

    /**
     * Get a TypeId parameter value.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @param index [in] - Array index. If parameter is not an array this shall be 0.
     * @return Parameter value.
     */
    public static long getTypeId(long typeId, int parameter, int index)
    {
        long val [] = new long [1];
        Kernel.GetInt64Parameter(typeId, parameter, index, val);
        return val[0];
    }

    /**
     * Get a InstanceId parameter value.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @param index [in] - Array index. If parameter is not an array this shall be 0.
     * @return Parameter value.
     */
    public static InstanceId getInstanceId(long typeId, int parameter, int index)
    {
        long hashVal[] = new long[1];
        String strVal[] = new String[1];

        Kernel.GetHashedIdParameter(typeId, parameter, index, hashVal, strVal);
        if (strVal[0] != null) {
            return new InstanceId(hashVal[0], strVal[0]);
        }
        else {
            return new InstanceId(hashVal[0]);
        }
    }

    /**
     * Get an EntityId parameter value.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @param index [in] - Array index. If parameter is not an array this shall be 0.
     * @return Parameter value.
     */
    public static EntityId getEntityId(long typeId, int parameter, int index)
    {
        long typeIdOut[] = new long [1];
        long instanceId[] = new long [1];
        String strVal[] = new String[1];
        Kernel.GetEntityIdParameter(typeId, parameter, index, typeIdOut, instanceId , strVal);
        if (strVal[0] != null) {
            return new EntityId(typeIdOut[0], new InstanceId(instanceId[0], strVal[0]));
        }
        else {
            return new EntityId(typeIdOut[0], new InstanceId(instanceId[0]));
        }
    }

    /**
     * Get a ChannelId parameter value.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @param index [in] - Array index. If parameter is not an array this shall be 0.
     * @return Parameter value.
     */
    public static ChannelId getChannelId(long typeId, int parameter, int index)
    {
        long hashVal[] = new long[1];
        String strVal[] = new String[1];

        Kernel.GetHashedIdParameter(typeId, parameter, index, hashVal, strVal);
        if (strVal[0] != null) {
            return new ChannelId(hashVal[0], strVal[0]);
        }
        else {
            return new ChannelId(hashVal[0]);
        }
    }

    /**
     * Get a HandlerId parameter value.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @param index [in] - Array index. If parameter is not an array this shall be 0.
     * @return Parameter value.
     */
    public static HandlerId getHandlerId(long typeId, int parameter, int index)
    {
        long hashVal[] = new long[1];
        String strVal[] = new String[1];

        Kernel.GetHashedIdParameter(typeId, parameter, index, hashVal, strVal);
        if (strVal[0] != null) {
            return new HandlerId(hashVal[0], strVal[0]);
        }
        else {
            return new HandlerId(hashVal[0]);
        }
    }


    /**
     * Get an Object parameter value.
     *
     * This method will return a smart pointer to a new copy of the parameter.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @param index [in] - Array index. If parameter is not an array this shall be 0.
     * @return Parameter value.
     */
    public static com.saabgroup.safir.dob.typesystem.Object getObject(long typeId, int parameter, int index)
    {
        java.nio.ByteBuffer blob [] = new java.nio.ByteBuffer[1];
        Kernel.GetObjectParameter(typeId, parameter, index, blob);
        return ObjectFactory.getInstance().createObject(blob[0]);
    }

    /**
     * Get a Binary parameter value.
     *
     * This method will retrieve a pointer to the binary data and the number of bytes the binary data contains.
     *
     * @param typeId [in] - TypeId of class.
     * @param parameter [in] - Index of parameter.
     * @param index [in] - Array index. If parameter is not an array this shall be 0.
     * @return Parameter value.
     */
    public static byte[] getBinary(long typeId, int parameter, int index)
    {
        java.nio.ByteBuffer val [] = new java.nio.ByteBuffer [1];
        Kernel.GetBinaryParameter(typeId, parameter, index, val);
        byte[] result = new byte [val[0].capacity()];
        val[0].clear(); //reset position
        val[0].get(result);
        return result;
    }
}
